R/rdeephaven/src/client.cpp

```diff
@@ -0,0 +1,233 @@
+
+
+
+class ClientOptionsWrapper {
+public:
+
+    ClientOptionsWrapper() {
+        internal_options = new deephaven::client::ClientOptions();
+        bool authentication_set = false;
+        bool session_set = false;
+    };
+
+    void setDefaultAuthentication() {
+        internal_options->setDefaultAuthentication();
+        authentication_set = true;
+    };
+
+    void setBasicAuthentication(const std::string &username, const std::string &password) {
+        internal_options->setBasicAuthentication(username, password);
+        authentication_set = true;
+    };
+
+    void setCustomAuthentication(const std::string &authenticationKey, const std::string &authenticationValue) {
+        internal_options->setCustomAuthentication(authenticationKey, authenticationValue);
+        authentication_set = true;
+    };
+    
+    void setSessionType(const std::string &sessionType) {
+        internal_options->setSessionType(sessionType);
+        session_set = true;
+    };
+
+    bool authentication_set;
+    bool session_set;
+
+private:
+
+    deephaven::client::ClientOptions* internal_options;
+    friend ClientWrapper* newClientWrapper(const std::string &target, const ClientOptionsWrapper &client_options);
+};
+
+
+
+class ClientWrapper {
+public:
+
+    /**
+     * Fetches a reference to a table named tableName on the server if it exists.
+     * @param tableName Name of the table to search for.
+     * @return TableHandle reference to the fetched table. 
+    */
+    TableHandleWrapper* openTable(std::string tableName) {
+        return new TableHandleWrapper(internal_tbl_hdl_mngr.fetchTable(tableName));
+    };
+
+    /**
+     * Runs a script on the server in the console language if a console was created.
+     * @param code String of the code to be executed on the server.
+    */
+    void runScript(std::string code) {
+        internal_tbl_hdl_mngr.runScript(code);
+    };
+
+    /**
+     * Checks for the existence of a table named tableName on the server.
+     * @param tableName Name of the table to search for.
+     * @return Boolean indicating whether tableName exists on the server or not.
+    */
+    bool checkForTable(std::string tableName) {
+        // we have to first fetchTable to check existence, fetchTable does not fail on its own, but .observe() will fail if table doesn't exist
+        deephaven::client::TableHandle table_handle = internal_tbl_hdl_mngr.fetchTable(tableName);
+        try {
+            table_handle.observe();
+        } catch(...) {
+            return false;
+        }
+        return true;
+    };
+
+    /**
+     * Allocates memory for an ArrowArrayStream C struct and returns a pointer to the new chunk of memory.
+     * Intended to be used to get a pointer to pass to Arrow's R library RecordBatchReader$export_to_c(ptr).
+    */
+    SEXP newArrowArrayStreamPtr() {
+        ArrowArrayStream* stream_ptr = new ArrowArrayStream();
+        return Rcpp::XPtr<ArrowArrayStream>(stream_ptr, true);
+    };
```

Comment:
As I noted elsewhere, this method has nothing to do with the client, so it logically does not belong here.  Can be fixed v2, or we can leave it like this.